### PR TITLE
B16 (API Changes)

### PR DESCRIPTION
**Pull request by vinniefalco**
* Make value optional in param-list
* Frame processing routines are member functions
* Fix on_headers called twice from basic_parser_v1
* Constrain parser_v1 constructor
* Improve first line serialization
* Add pause option to on_headers interface
* Refactor base_parser_v1 callback traits:
* Refine Parser concept
* Relax ForwardIterator requirements in FieldSequence

API Changes:

* Rename mask_buffer_size to write_buffer_size
* Make auto_fragment a boolean option

### CODE REVIEW

File: include/beast/http/detail/rfc7230.hpp [hunk 1]
```diff
@@ -233,6 +233,19 @@ skip_ows(FwdIt& it, FwdIt const& end)
...
+void
+skip_token(FwdIt& it, FwdIt const& end)
+{
+    while(it != end)
+    {
+        auto const c = *it;
+        if(! is_tchar(c))
+            break;
+        ++it;
+    }
```
Comment (miguelportilla) [inline include/beast/http/detail/rfc7230.hpp]: alternate suggestion:

while(it != end && is_tchar(*it))
    ++it;

**Comment by vinniefalco** [comment]
The loss of code coverage is disappointing

File: doc/design.qbk [hunk 2]
```diff
@@ -5,38 +5,46 @@
...
-* Symmetric interfaces (client and server the same, or close to it).
+* Mimic Boost.Asio; familiarity breeds confidence.
 
-* Emulate Boost.Asio interfaces as much as possible, since Asio is
-  proven and it is familiar to users.
+* Role-symmetric interfaces; client and server the same (or close to it).
 
-* Let library users make the important decisions such as how to
-  allocate memory or how to leverage flow control.
+* Leave important decisions to the user, such as
```
Comment (nbougalis) [inline doc/design.qbk]: Reads weird. Consider:

> Leave important decisions, such as memory allocation and flow control, to the user.

File: doc/master.qbk [hunk 3]
```diff
@@ -49,8 +49,7 @@ provides implementations of the HTTP and WebSocket protocols.
   [[
     [link beast.overview Overview]
   ][
-    An overview of features, requirements, and credits, plus
-    rationale and design information.
+    An introduction wsth features, requirements, and credits.
```
Comment (nbougalis) [inline doc/master.qbk]: Typo.

File: doc/types/Parser.qbk [hunk 4]
```diff
@@ -27,18 +30,18 @@ In this table:
     [`a.complete()`]
     [`bool`]
     [
-        Returns `true` when a complete HTTP/1 message has been parsed.
+        Returns `true` when the parser is complete.
```
Comment (nbougalis) [inline doc/types/Parser.qbk]: "the parser" should be "parsing".

File: include/beast/http/basic_parser_v1.hpp [hunk 5]
```diff
@@ -153,14 +181,17 @@ static std::uint64_t constexpr no_content_length =
...
 
-    @li *0* The parser should expect a body
+    @li `normal` The parser should expect a body
 
-    @li *1* The parser should skip the body. For example, this is
+    @li `skip` The parser should skip the body. For example, this is
         used when sending a response to a HEAD request.
 
-    @li *2* The parser should skip ths body, this is an
+    @li `upgrade` The parser should skip ths body, this is an
```
Comment (nbougalis) [inline include/beast/http/basic_parser_v1.hpp]: Typo *ths*

Comment (nbougalis) [inline include/beast/http/basic_parser_v1.hpp]: Also the comma would better as a semicolon.

File: include/beast/http/rfc7230.hpp [hunk 6]
```diff
@@ -15,15 +15,14 @@ namespace http {
 
 /** A list of parameters in a HTTP extension field value.
 
-    This container allows iteration of the parameter list
-    in a HTTP extension. The parameter list is a series
-    of "name = value" pairs with each pair starting with
-    a semicolon.
+    This container allows iteration of the parameter list in a HTTP
+    extension. The parameter list is a series of name/value" pairs
```
Comment (nbougalis) [inline include/beast/http/rfc7230.hpp]: Rogue "

File: include/beast/websocket/impl/stream.ipp [hunk 7]
```diff
@@ -793,82 +642,157 @@ write_frame(bool fin, ConstBufferSequence const& buffers)
...
     fh.mask = role_ == detail::role_type::client;
-    if(fh.mask)
-        fh.key = maskgen_();
-    detail::fh_streambuf fh_buf;
-    detail::write<static_streambuf>(fh_buf, fh);
-    if(! fh.mask)
+    auto remain = buffer_size(buffers);
+    if(compress)
     {
-        // send header and payload
```
Comment (nbougalis) [inline include/beast/websocket/impl/stream.ipp]: Consider a comment to indicate the block is intentionally empty.

File: test/websocket/stream.cpp [hunk 8]
```diff
@@ -911,21 +911,23 @@ class stream_test
                 ws.set_option(pong_callback{});
 
                 // send auto fragmented message
-                ws.set_option(auto_fragment_size(3));
-                ws.write(sbuf("Hello"));
+                ws.set_option(auto_fragment{true});
+                ws.set_option(write_buffer_size{8});
+                ws.write(sbuf("Now is the time for all good men"));
```
Comment (nbougalis) [inline test/websocket/stream.cpp]: Consider also testing with messages that are 7, 8 and 9 bytes.

File: doc/types/Parser.qbk (hunk 4 above)
Comment (vinniefalco) [inline doc/types/Parser.qbk]: Will fix

File: include/beast/http/basic_parser_v1.hpp (hunk 5 above)
Comment (vinniefalco) [inline include/beast/http/basic_parser_v1.hpp]: Will fix

File: include/beast/websocket/impl/stream.ipp (hunk 7 above)
Comment (vinniefalco) [inline include/beast/websocket/impl/stream.ipp]: I can't find the code anymore

**Comment by vinniefalco** [comment]
Added "Cmake install and VS2015 support"

File: doc/types/FieldSequence.qbk [hunk 9]
```diff
@@ -23,14 +23,14 @@ In this table:
     [`X::value_type`]
     []
     [
-        A type that meets the requirements of `Field`.
+        A type that meets the requirements of @b Field.
```
Comment (vinniefalco) [inline doc/types/FieldSequence.qbk]: Should be `[*ForwardIterator]` to get bold, the `@b` is Javadoc markdown.

File: include/beast/http/rfc7230.hpp (hunk 6 above)
Comment (vinniefalco) [inline include/beast/http/rfc7230.hpp]: will fix

File: include/beast/websocket/impl/stream.ipp (hunk 7 above)
Comment (vinniefalco) [inline include/beast/websocket/impl/stream.ipp]: Will fix, good find.

File: test/websocket/stream.cpp (hunk 8 above)
Comment (vinniefalco) [inline test/websocket/stream.cpp]: This is still a little bit in flux, I'm going to wait until I have finished the permessage-deflate extension and then revise all the tests, and try to get coverage up.

File: include/beast/http/detail/rfc7230.hpp [hunk 10]
```diff
@@ -318,20 +324,23 @@ increment()
     {
         ++it;
         if(it == end)
-            return err();
+            break;
         if(! detail::is_tchar(*it))
             break;
     }
```
Comment (miguelportilla) [inline include/beast/http/detail/rfc7230.hpp]: @ximinez  Good suggestion, as long as `it` is incremented before the call.

Or

while(++it != end && is_tchar(*it));

File: include/beast/http/detail/rfc7230.hpp [hunk 11]
```diff
@@ -318,20 +319,23 @@ increment()
...
     auto const p1 = it;
+    v.first = { &*p0, static_cast<std::size_t>(p1 - p0) };
     detail::skip_ows(it, end);
     if(it == end)
-        return err();
+        return;
```
Comment (vinniefalco) [inline include/beast/http/detail/rfc7230.hpp]: Why would I do that?

File: include/beast/http/detail/rfc7230.hpp (hunk 10 above)
Comment (vinniefalco) [inline include/beast/http/detail/rfc7230.hpp]: Are you saying to replace
```
    for(;;)
    {
        ++it;
        if(it == end)
            break;
        if(! detail::is_tchar(*it))
            break;
    }
```
with
```
while(++it != end && is_tchar(*it));
```
?

File: doc/websocket.qbk [hunk 12]
```diff
@@ -302,13 +302,14 @@ in the error code [link beast.ref.websocket__error `error::closed`] being
...
-gives callers control over the size of these frames:
+To ensure timely delivery of control frames, large messages can be broken up
+into smaller sized frames. The automatic fragment option turns on this
+feature, and the write buffer size option determines the maximum size of
+the fragments:
 ```
     ...
-    ws.set_option(beast::websocket::auto_fragment_size{8192});
+    ws.set_option(beast::websocket::auto_fragment{true};
+    ws.set_option(beast::websocket::write_buffer_size{16384};
```
Comment (vinniefalco) [inline doc/websocket.qbk]: Will fix

File: include/beast/http/detail/rfc7230.hpp (hunk 11 above)
Comment (miguelportilla) [inline include/beast/http/detail/rfc7230.hpp]: Now some cases use `return;` and some `return err();` which is a bit misleading in the sense that something will return.

File: include/beast/http/detail/rfc7230.hpp [hunk 13]
```diff
@@ -371,7 +379,6 @@ increment()
             if(! detail::is_tchar(*it))
                 break;
         }
```
Comment (miguelportilla) [inline include/beast/http/detail/rfc7230.hpp]: Same as above. At the very least `it++` should be `++it`

File: include/beast/http/detail/rfc7230.hpp (hunk 10 above)
Comment (vinniefalco) [inline include/beast/http/detail/rfc7230.hpp]: `skip_token(++it, end);` looks good, will fix

File: include/beast/http/detail/rfc7230.hpp (hunk 13 above)
Comment (vinniefalco) [inline include/beast/http/detail/rfc7230.hpp]: Will fix

File: include/beast/http/detail/rfc7230.hpp (hunk 11 above)
Comment (miguelportilla) [inline include/beast/http/detail/rfc7230.hpp]: I see what you're going for but I think `{err(); return;}` is clearer.
Considering snippet `err(); return` tells me both functions return void. Where `return err();` does not. IMHO

File: test/http/rfc7230.cpp [hunk 14]
```diff
@@ -43,8 +43,11 @@ class rfc7230_test : public beast::unit_test::suite
         {
             s.push_back(';');
             s.append(str(p.first));
-            s.push_back('=');
-            s.append(str(p.second));
+            if(! p.second.empty())
```
Comment (miguelportilla) [inline test/http/rfc7230.cpp]: Suggestion: It may be beneficial to call s.reserve before the loop.

File: doc/types/FieldSequence.qbk (hunk 9 above)
Comment (vinniefalco) [inline doc/types/FieldSequence.qbk]: will fix

File: doc/types/Parser.qbk [hunk 15]
```diff
@@ -7,17 +7,20 @@
...
 
-* `X` denotes a type meeting the requirements of [*`Parser`].
+* `X` denotes a type meeting the requirements of [*Parser].
 
 * `a` denotes a value of type `X`.
 
-* `b` is a value meeting the requirements of [@http://www.boost.org/doc/libs/1_61_0/doc/html/boost_asio/reference/ConstBufferSequence.html [*`ConstBufferSequence`]].
+* `b` is a value meeting the requirements of __ConstBufferSequence__.
 
 * `ec` is a value of type [link beast.ref.error_code `error_code&`].
```
Comment (vinniefalco) [inline doc/types/Parser.qbk]: I believe this is correct as written.

File: include/beast/http/basic_parser_v1.hpp [hunk 16]
```diff
@@ -186,7 +186,7 @@ static std::uint64_t constexpr no_content_length =
     @li `skip` The parser should skip the body. For example, this is
         used when sending a response to a HEAD request.
 
-    @li `upgrade` The parser should skip ths body, this is an
+    @li `upgrade` The parser should skip this body; its an
```
Comment (vinniefalco) [inline include/beast/http/basic_parser_v1.hpp]: Will fix

File: include/beast/http/basic_parser_v1.hpp [hunk 17]
```diff
@@ -153,14 +181,17 @@ static std::uint64_t constexpr no_content_length =
...
-    @li *1* The parser should skip the body. For example, this is
+    @li `skip` The parser should skip the body. For example, this is
         used when sending a response to a HEAD request.
 
-    @li *2* The parser should skip ths body, this is an
+    @li `upgrade` The parser should skip ths body, this is an
         upgrade to a different protocol.
 
+    @li `pause` The parser should preserve the parsing state and
+        return control to the caller.
```
Comment (vinniefalco) [inline include/beast/http/basic_parser_v1.hpp]: Will fix

File: test/http/rfc7230.cpp (hunk 14 above)
Comment (vinniefalco) [inline test/http/rfc7230.cpp]: I have a hard time getting excited about a microoptimization in unit test code...

Comment (miguelportilla) [inline test/http/rfc7230.cpp]: Agreed! I forgot this was test.

File: include/beast/websocket/detail/mask.hpp [hunk 18]
```diff
@@ -67,7 +67,9 @@ maskgen_t<_>::rekey()
     g_.seed(ss);
 }
 
-using maskgen = maskgen_t<std::mt19937>;
+// VFALCO NOTE This generator has 5KB of state!
+//using maskgen = maskgen_t<std::mt19937>;
```
Comment (miguelportilla) [inline include/beast/websocket/detail/mask.hpp]: Super nit-pick: missing space

File: include/beast/websocket/impl/stream.ipp [hunk 19]
```diff
@@ -657,82 +642,158 @@ write_frame(bool fin, ConstBufferSequence const& buffers)
...
+    copy buffers to write_buffer
+    apply mask to write_buffer
+    write frame header and possibly full write_buffer in a single call
+    loop:
+        copy buffers to write_buffer
+        apply mask to write_buffer
+        write write_buffer in a single call
+else
+        write frame header, buffers as one frame
+*/
```
Comment (miguelportilla) [inline include/beast/websocket/impl/stream.ipp]: Was this comment intentionally left?

File: include/beast/http/detail/rfc7230.hpp (hunk 11 above)
Comment (vinniefalco) [inline include/beast/http/detail/rfc7230.hpp]: You should get used to `void` returns, they are already possible in the language and soon they will be treatable as first class types, see http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0146r1.html

File: include/beast/websocket/detail/stream_base.hpp [hunk 20]
```diff
@@ -115,6 +131,287 @@ struct stream_base
...
+    switch(rd_fh_.op)
+    {
+    case opcode::binary:
+    case opcode::text:
+        if(rd_cont_)
+        {
+            // new data frame when continuation expected
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.rsv1 || rd_fh_.rsv2 || rd_fh_.rsv3)
```
Comment (vinniefalco) [inline include/beast/websocket/detail/stream_base.hpp]: Hmm the current version is a bit more clear

File: include/beast/websocket/detail/stream_base.hpp [hunk 21]
```diff
@@ -115,6 +131,287 @@ struct stream_base
...
+        }
+        break;
+
+    case opcode::cont:
+        if(! rd_cont_)
+        {
+            // continuation without an active message
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.rsv1 || rd_fh_.rsv2 || rd_fh_.rsv3)
```
Comment (vinniefalco) [inline include/beast/websocket/detail/stream_base.hpp]: Hmm the current version is a bit more clear

File: include/beast/websocket/detail/stream_base.hpp [hunk 22]
```diff
@@ -115,6 +131,287 @@ struct stream_base
...
+        {
+            // fragmented control message
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.len > 125)
+        {
+            // invalid length for control message
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.rsv1 || rd_fh_.rsv2 || rd_fh_.rsv3)
```
Comment (vinniefalco) [inline include/beast/websocket/detail/stream_base.hpp]: Hmm the current version is a bit more clear. Also, this line will change with permessage-deflate (which uses some of the now-reserved bits).

File: include/beast/websocket/impl/stream.ipp (hunk 19 above)
Comment (vinniefalco) [inline include/beast/websocket/impl/stream.ipp]: Yes its part of some implementation notes

**Comment by vinniefalco** [comment]
Doc preview is here:
http://vinniefalco.github.io/beast

**Comment by vinniefalco** [comment]
Addressed all comments, the changes are individually marked

File: test/websocket/stream.cpp (hunk 8 above)
Comment (miguelportilla) [inline test/websocket/stream.cpp]: This loop breaks on the first iteration, even if `n` is less than 199.

Comment (vinniefalco) [inline test/websocket/stream.cpp]: That's a major bug... those loops should start from 0.

**Comment by miguelportilla** [comment]
overview.qbk misspells Christophers last name. should be Kohlhoff not Kohloff

**Pull request merged**